opt: track project equivalencies when inlining projections into filters

mgartner opened the pull request:
#### opt: add ProjectPrivate

ProjecExprs now have a ProjectPrivate field. For now it only contains
the passthrough columns. Future commits will add additional fields.

Release note: None

#### opt: add EquivGroups to ProjectPrivate

EquivGroups have been added to ProjectPrivate. Rules that inline
projections into filters populate this field with equivalencies that
would otherwise be lost in the filter push down. For example,
consider...

TODO: Add example.

Fixes #134816

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147512)
<!-- Reviewable:end -->

mgartner committed:
opt: add ProjectPrivate

ProjecExprs now have a ProjectPrivate field. For now it only contains
the passthrough columns. Future commits will add additional fields.

Release note: None

mgartner committed:
opt: add EquivGroups to ProjectPrivate

EquivGroups have been added to ProjectPrivate. Rules that inline
projections into filters populate this field with equivalencies that
would otherwise be lost in the filter push down. For example,
consider...

TODO: Add example.

Fixes #134816

Release note: None

mgartner commented:
In the latest iteration I've added a `ProjectPrivate` which contains passthrough columns and equivalency groups. It's a pretty big change and we lose the nice, easy-to-use `Passthrough ColSet` as the project's private field.

This is too complex to backport, so if we go ahead with this, I think we should disable the assertion triggered in #134816 to silence future failures on release branches.

I'm a bit hesitant about this approach because I worry that it makes it easy to duplicate a ProjectPrivate in a rule in which the new Project should not have the equivalency groups of the matched Project. Is understanding this new field a subtle detail that is more demanding than other expressions or rules?

rytaft reviewed (commented):
I agree that we shouldn't backport this, but it seems like a reasonable, pretty clean approach to the problem. I wonder if there are any cases where this also unlocks optimization opportunities we didn't have before?

> the new Project should not have the equivalency groups of the matched Project

Can you think of an example where this would be the case?

> Reviewed 47 of 47 files at r2, 8 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147512#-:-OS4vsC8-CdvZfYQDax7:bbh9q0v)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`-- commits` line 17 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147512#-OS4vOcw8JkMkBWQ7urG:-OS4vOcw8JkMkBWQ7urH:b-ovjwuh):*
Reminder to fill this in :) 
___
*[`pkg/sql/opt/norm/set_funcs.go` line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147512#-OS4sRMl0dannKKmDvV4:-OS4sRMl0dannKKmDvV5:b9o5ej) ([raw file](https://github.com/cockroachdb/cockroach/blob/52b13c9c4fe8f9ada5f0b40e804312e44beec6db/pkg/sql/opt/norm/set_funcs.go#L55)):*
> ```Go
> }
> 
> // ProjectPassthroughRight returns a ColSet that contains the columns that can
> ```

nit: update comment (here and above)
___
*[`pkg/sql/opt/memo/interner.go` line 468 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147512#-OS4s_455OGsr5NFeJjM:-OS4s_46AlMxAuMu3GL6:bj79y8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/52b13c9c4fe8f9ada5f0b40e804312e44beec6db/pkg/sql/opt/memo/interner.go#L468)):*
> ```Go
> }
> 
> func (h *hasher) HashEquivGroups(equiv props.EquivGroups) {
> ```

This should be part of the second commit (the change here and below)


<!-- Sent from Reviewable.io -->
